### PR TITLE
array: generalize S.head, S.last, S.tail, and S.init

### LINKED ISSUE #570: Update head and tail to work for all Foldable
Currently Head and Tail only work for `Array`, but it is possible to implement them for all `Foldable`.

### PR DESCRIPTION
**Pull request by syves**
Closes #570

Initial attempt to generalize head, tail, last and init functions for Foldables.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,7 +1,4 @@
 /*    #######
-   ####     ####
- ####   ###   ####
-#####   ###########   sanctuary
```
Comment (davidchambers) [inline index.js]: Please reinstate these three lines. :laughing:

File: index.js [hunk 2]
```diff
@@ -3073,38 +3087,78 @@
   //.
   //. > S.last ([])
   //. Nothing
+  //.
+  //. > S.last (Cons ('baz') (Cons ('foo') (Nil)))
+  //. S.Just ('foo')
```
Comment (davidchambers) [inline index.js]: I'd prefer to see `S.last (Cons ('foo') (Cons ('bar') (Cons ('baz') (Nil))))` here. Deviating from the customary _foo_, _bar_, _baz_ order could lead to confusion.

File: index.js [hunk 3]
```diff
@@ -3073,38 +3087,78 @@
...
+  //. Foldable contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.tail ([1, 2, 3])
   //. Just ([2, 3])
   //.
   //. > S.tail ([])
+  //.
+  //. > S.tail (Cons ('baz') (Cons ('foo') (Nil)))
+  //. Just (Cons ('foo') (Nil))
```
Comment (davidchambers) [inline index.js]: Let's stick with _foo_, _bar_, _baz_ order here as well.

File: index.js [hunk 4]
```diff
@@ -3073,38 +3087,78 @@
...
+  //
+  //. > S.tail (Nil)
   //. Nothing
   //. ```
+
+  function tail(foldable) {
+    var counter = 0
```
Comment (davidchambers) [inline index.js]: Is a counter necessary? Could you infer whether or not you're processing the first element from the value of the accumulator?

File: test/head.js [hunk 5]
```diff
@@ -1,19 +1,22 @@
...
-  eq (S.show (S.head)) ('head :: Array a -> Maybe a');
+  eq (S.show (S.head)) ('head :: Foldable f => f a -> Maybe a');
 
   eq (S.head ([])) (S.Nothing);
   eq (S.head (['foo'])) (S.Just ('foo'));
   eq (S.head (['foo', 'bar'])) (S.Just ('foo'));
   eq (S.head (['foo', 'bar', 'baz'])) (S.Just ('foo'));
 
+  eq (S.head (Cons ('foo') (Cons ('bar') (Cons ('baz') (Nil))))) (S.Just ('foo'));
+  eq (S.head (Nil)) (S.Nothing);
```
Comment (davidchambers) [inline test/head.js]: Let's include four list tests in the same order as the array tests.

File: test/init.js [hunk 6]
```diff
@@ -9,11 +9,11 @@ test ('init', () => {
...
   eq (typeof S.init) ('function');
   eq (S.init.length) (1);
-  eq (S.show (S.init)) ('init :: Array a -> Maybe (Array a)');
+  eq (S.show (S.init)) ('init :: (Foldable f, Applicative f) => f a -> Maybe (f a)');
 
   eq (S.init ([])) (S.Nothing);
   eq (S.init (['foo'])) (S.Just ([]));
   eq (S.init (['foo', 'bar'])) (S.Just (['foo']));
-  eq (S.init (['foo', 'bar', 'baz'])) (S.Just (['foo', 'bar']));
+  //eq (S.init (['foo', 'bar', 'baz'])) (S.Just (['foo', 'bar']));
```
Comment (davidchambers) [inline test/init.js]: Please remove this line, and add four list tests.

File: test/tail.js [hunk 7]
```diff
@@ -1,19 +1,22 @@
...
+  eq (S.show (S.tail)) ('tail :: (Foldable f, Applicative f) => f a -> Maybe (f a)');
 
   eq (S.tail ([])) (S.Nothing);
   eq (S.tail (['foo'])) (S.Just ([]));
   eq (S.tail (['foo', 'bar'])) (S.Just (['bar']));
   eq (S.tail (['foo', 'bar', 'baz'])) (S.Just (['bar', 'baz']));
 
+  eq (S.tail (Cons ('baz') (Cons ('foo') (Nil)))) (S.Just (Cons ('foo') (Nil)));
+  eq (S.tail (Nil)) (S.Nothing);
 });
```
Comment (davidchambers) [inline test/tail.js]: Please insert a blank line after the last assertion.

File: index.js [hunk 8]
```diff
@@ -3073,38 +3087,78 @@
...
   _.last = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
     impl: last
   };
 
-  //# tail :: Array a -> Maybe (Array a)
+  //# tail :: (Foldable f, Applicative f) => f a -> Maybe (f a)
```
Comment (davidchambers) [inline index.js]: Could you justify the Applicative constraint?

File: index.js (hunk 2 above)
Comment (davidchambers) [reply index.js]: The expected output should now be `Just ('baz')`, I believe. You could run `npm run doctest` to check.

File: index.js [hunk 9]
```diff
@@ -3073,52 +3090,121 @@
...
-  //. array contains at least one element; Nothing otherwise.
+  //. Returns a Just of all but the first of the given Foldable if the
+  //. Foldable contains at least one element; Nothing otherwise.
   //.
   //. ```javascript
   //. > S.tail ([1, 2, 3])
   //. Just ([2, 3])
   //.
   //. > S.tail ([])
+  //.
```
Comment (davidchambers) [inline index.js]: The doctest above lost its output line.

File: index.js [hunk 10]
```diff
@@ -3073,52 +3090,121 @@
...
+  //. > S.init (Cons ('bar') (Nil))
+  //. Just (Nil)
+  //.
+  //. > S.init (Nil)
+  //. Nothing
   //. ```
-  function init(xs) {
-    return xs.length > 0 ? Just (xs.slice (0, -1)) : Nothing;
+
+//. ```
```
Comment (davidchambers) [inline index.js]: This line can be deleted, I think.

File: index.js (hunk 2 above)
Comment (syves) [reply index.js]: 🆗 ok

File: index.js (hunk 4 above)
Comment (syves) [reply index.js]: 🆗

File: test/head.js (hunk 5 above)
Comment (syves) [reply test/head.js]: 🆗

File: test/init.js (hunk 6 above)
Comment (syves) [reply test/init.js]: 🆗

File: index.js (hunk 8 above)
Comment (syves) [reply index.js]: In order to build a new foldable with the right order I need to use 'append' which requires that f be an Applicative and Semigroup. I'm now a bit confused about why it is working without both being required.

File: index.js (hunk 3 above)
Comment (syves) [reply index.js]: 🆗

File: test/tail.js (hunk 7 above)
Comment (syves) [reply test/tail.js]: This file has been updated.

File: index.js (hunk 9 above)
Comment (syves) [reply index.js]: 🆗

File: index.js (hunk 8 above)
Comment (davidchambers) [reply index.js]: > In order to build a new foldable with the right order I need to use 'append' which requires that f be an Applicative and Semigroup.

Good point! It looks as though you're also using [`Z.empty`][1], so there may be other constraints to consider.

> I'm now a bit confused about why it is working without both being required.

Type classes need not be specified in order for a function to work. In fact, the only reason to specify a function's types and type-class constraints is to ensure that *invalid* applications of the function result in errors with descriptive messages. For example:

```javascript
S.concat (123) (456);
// ! TypeError: Type-class constraint violation
//
//   concat :: Semigroup a => a -> a -> a
//             ^^^^^^^^^^^    ^
//                            1
//
//   1)  123 :: Number, FiniteNumber, NonZeroFiniteNumber, Integer, NonNegativeInteger, ValidNumber
//
//   ‘concat’ requires ‘a’ to satisfy the Semigroup type-class constraint; the value at position 1 does not.
//
//   See https://github.com/sanctuary-js/sanctuary-type-classes/tree/v10.0.0#Semigroup for information about the sanctuary-type-classes/Semigroup type class.
```

If we replace `{a: [Z.Semigroup]}` with `{}` and try again, the error is less helpful:

```javascript
S.concat (123) (456);
// ! TypeError: Semigroup.methods.concat(...) is not a function
```


[1]: https://github.com/sanctuary-js/sanctuary-type-classes/tree/v10.0.0#empty

File: index.js (hunk 3 above)
Comment (davidchambers) [reply index.js]: This output looks incorrect.

File: test/last.js [hunk 11]
```diff
@@ -1,19 +1,26 @@
...
   eq (S.last (['foo'])) (S.Just ('foo'));
   eq (S.last (['foo', 'bar'])) (S.Just ('bar'));
   eq (S.last (['foo', 'bar', 'baz'])) (S.Just ('baz'));
 
+  eq (S.last (Nil)) (S.Nothing);
+  eq (S.last (Cons ('foo') (Nil))) (S.Just ('foo'))
+  eq (S.last (Cons ('foo') (Cons ('bar') (Nil))))
+  (S.Just ('bar'));
+  eq (S.last (Cons ('foo') (Cons ('bar') (Cons ('baz') (Nil)))))
+     (S.Just ('baz'));
```
Comment (davidchambers) [inline test/last.js]: Don't worry about line length in the test files. These assertions would read better as one-liners.

File: index.js (hunk 3 above)
Comment (syves) [reply index.js]: 🆗

File: index.js (hunk 8 above)
Comment (syves) [reply index.js]: Verbose messages are much more helpful.

File: test/last.js (hunk 11 above)
Comment (syves) [reply test/last.js]: 🆗

**Comment by syves** [comment]
I've addressed the above changes, but for some reason doctest is not running.

> 20 error code ELIFECYCLE
> 21 error errno 1
> 22 error sanctuary@1.0.0 doctest: `sanctuary-doctest`
> 22 error Exit status 1
> 23 error Failed at the sanctuary@1.0.0 doctest script.

File: test/last.js (hunk 11 above)
Comment (davidchambers) [reply test/last.js]: It looks as though you deleted some of the assertions. I was suggesting that you remove the line breaks rather than the whole lines. ;)

File: index.js (hunk 8 above)
Comment (davidchambers) [reply index.js]: > It looks as though you're also using [`Z.empty`][1], so there may be other constraints to consider.

Have you given this some thought?


[1]: https://github.com/sanctuary-js/sanctuary-type-classes/tree/v10.0.0#empty

Comment (syves) [reply index.js]: Now I have :)

Comment (syves) [reply index.js]: 🆗

File: index.js [hunk 12]
```diff
@@ -3044,81 +3044,151 @@
...
+  //. > S.head (Cons ('foo') (Cons ('bar') (Nil)))
+  //. S.Just ('foo')
+  //.
+  //. > S.head (Nil)
+  //. S.Nothing
   //. ```
+
+  function head(foldable) {
+    return Z.reduce (
+      function(acc, a) {
```
Comment (davidchambers) [inline index.js]: For consistency with other implementations, let's name these parameters `maybe` and `x`.

File: index.js [hunk 13]
```diff
@@ -3044,81 +3044,151 @@
     impl: at
   };
 
-  //# head :: Array a -> Maybe a
+  //# head :: Foldable f => f a -> Maybe a
   //.
-  //. Returns Just the first element of the given array if the array contains
-  //. at least one element; Nothing otherwise.
+  //. Returns Just the first element of the given collection if the foldable
+  //. contains at least one element; Nothing otherwise.
```
Comment (davidchambers) [inline index.js]: We avoid the term _collection_ because people tend not to think of values of types such as `Identity a`, `Maybe a`, and `Either a b` as collections.

I think we should also avoid _foldable_. Referring to a type class works in this case, but this approach would be problematic for functions which place multiple constraints on a single type variable. Let's avoid referring to _applicative semigroups_ and _foldable applicative monoids_. :joy:

_Structure_ is the best term we have considered thus far.

File: index.js (hunk 12 above)
Comment (syves) [reply index.js]: :ok done

File: index.js (hunk 13 above)
Comment (syves) [reply index.js]: 🆗

File: index.js (hunk 12 above)
Comment (syves) [reply index.js]: applied to all 4

File: index.js (hunk 13 above)
Comment (syves) [reply index.js]: I understand this now.

File: index.js [hunk 14]
```diff
@@ -3044,81 +3044,151 @@
...
   _.last = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
     impl: last
   };
 
-  //# tail :: Array a -> Maybe (Array a)
+  //# tail :: (Applicative f, Foldable f, Monoid f) => f a -> Maybe (f a)
```
Comment (davidchambers) [inline index.js]:
```suggestion
  //# tail :: (Applicative f, Foldable f, Monoid (f a)) => f a -> Maybe (f a)
```

It's not yet possible to precisely describe this function's type using `def`, but the textual signature can be precise. The third constraint should be `Monoid (f a)` rather than `Monoid f`. `Maybe a`, for example, satisfies Monoid only if `a` satisfies Semigroup ([source][1]). The question we want to ask is not whether `Maybe` satisfies Monoid, but whether `Maybe String` satisfies Monoid. Does this make sense?


[1]: https://github.com/sanctuary-js/sanctuary-maybe/blob/v1.1.0/index.js#L117

File: index.js [hunk 15]
```diff
@@ -3044,81 +3044,151 @@
...
+      function(maybe, x) {
+        return maybe.isNothing ?
+               Just (Z.empty (foldable.constructor)) :
+               Z.map (append (x), maybe);
+      },
+      Nothing,
+      foldable
+    );
+  }
+
```
Comment (davidchambers) [inline index.js]: The empty lines before and after the implementations can be removed.

File: index.js [hunk 16]
```diff
@@ -3044,81 +3044,151 @@
...
   //. Just (1)
   //.
   //. > S.head ([])
   //. Nothing
+  //.
+  //. > S.head (Cons ('foo') (Cons ('bar') (Nil)))
+  //. S.Just ('foo')
+  //.
+  //. > S.head (Nil)
+  //. S.Nothing
```
Comment (davidchambers) [inline index.js]: For the expected output we like to use `Nothing` and `Just` rather than `S.Nothing` and `S.Just`. This makes the doctests for [`S.match`][1] and several other functions less noisy.


[1]: https://sanctuary.js.org/#match

File: index.js (hunk 14 above)
Comment (syves) [reply index.js]: yeah that makes sense

File: index.js (hunk 15 above)
Comment (syves) [reply index.js]: 🆗

File: index.js (hunk 16 above)
Comment (syves) [reply index.js]: 🆗

File: index.js [hunk 17]
```diff
@@ -3044,81 +3044,150 @@
...
+  //. > S.tail (Nil)
+  //. Nothing
   //. ```
+
+  function tail(foldable) {
+    return Z.reduce (
+      function(maybe, x) {
+        return maybe.isNothing ?
+               Just (Z.empty (foldable.constructor)) :
+               Z.map (append (x), maybe);
```
Comment (davidchambers) [inline index.js]: Something doesn't feel right about using `Z.map` in conjunction with the `isNothing` check. If we're to keep the `isNothing` check we could safely use `maybe.value` in the Just case:

```javascript
return maybe.isNothing ?
       Just (Pair (Z.empty (foldable.constructor)) (x)) :
       Just (Pair (Z.append (maybe.value.snd, maybe.value.fst)) (x));
```

This could be improved by removing the common code from the branches:

```javascript
return Just (maybe.isNothing ?
             Pair (Z.empty (foldable.constructor)) (x) :
             Pair (Z.append (maybe.value.snd, maybe.value.fst)) (x));
```

```javascript
return Just (Pair (maybe.isNothing ?
                   Z.empty (foldable.constructor) :
                   Z.append (maybe.value.snd, maybe.value.fst))
                  (x));
```

We could even remove the explicit conditional logic by using `Z.reduce`:

```javascript
return Just (Pair (Z.reduce (function(_, pair) { return Z.append (pair.snd, pair.fst); },
                             Z.empty (foldable.constructor),
                             maybe))
                  (x));
```

The problem with this approach is that `Z.empty (foldable.constructor)` would be evaluated once for each element of the structure. Can you think of a solution to this problem?

File: index.js (hunk 13 above)
Comment (davidchambers) [reply index.js]: > Returns Just the first element of the given structure if *it* contains at least one element; Nothing otherwise.

This is ambiguous: _it_ could refer to _the given structure_ or to _the first element of the given structure_. Let's use _the structure_ in place of _it_ to be clear, even though it's repetitive.

File: index.js [hunk 18]
```diff
@@ -3044,81 +3044,150 @@
...
   //. > S.head ([])
   //. Nothing
+  //.
+  //. > S.head (Cons ('foo') (Cons ('bar') (Nil)))
+  //. Just ('foo')
+  //.
+  //. > S.head (Nil)
+  //. Nothing
   //. ```
+
```
Comment (davidchambers) [inline index.js]: Let's remove the empty lines preceding the implementations.

File: index.js (hunk 13 above)
Comment (syves) [reply index.js]: 🆗

File: index.js (hunk 14 above)
Comment (syves) [reply index.js]: 🆗

File: index.js (hunk 18 above)
Comment (syves) [reply index.js]: 🆗

File: index.js (hunk 17 above)
Comment (davidchambers) [reply index.js]: The comment above applies to `init` rather than to `tail`. I'm sorry for causing confusion.

Comment (syves) [reply index.js]: I stored the empty in a var.

**Comment by davidchambers** [comment]
Let's complete this pull request together. I will push some commits to your branch. :)

File: index.js [hunk 19]
```diff
@@ -3063,81 +3063,123 @@
...
     impl: last
   };
 
-  //# tail :: Array a -> Maybe (Array a)
+  //# tail :: (Applicative f, Foldable f, Monoid (f a)) => f a -> Maybe (f a)
   //.
-  //. Returns Just all but the first of the given array's elements if the
-  //. array contains at least one element; Nothing otherwise.
+  //. Returns a Just of all but the first of the given structure if the
+  //. structure contains at least one element; Nothing otherwise.
```
Comment (davidchambers) [inline index.js]: We lost the word _elements_ from this sentence.

The change from _Just_ to _a Just of_ should be discussed in a separate pull request or issue. If we're to make this change we should do so for all descriptions, not just those we happen to be editing.

File: index.js [hunk 20]
```diff
@@ -3063,81 +3063,123 @@
...
+  //. > S.tail (Nil)
+  //. Nothing
   //. ```
+  function tail(foldable) {
+    return Z.reduce (function(m, x) {
+      return m.isNothing ?
+             Just (Z.empty (foldable.constructor)) :
+             Z.map (append (x), m);
+    }, Nothing, foldable);
+  }
```
Comment (davidchambers) [inline index.js]: What do you think of this change, @syves?

```diff
   function tail(foldable) {
+    var empty = Z.empty (foldable.constructor);
     return Z.reduce (function(m, x) {
-      return m.isNothing ?
-             Just (Z.empty (foldable.constructor)) :
-             Z.map (append (x), m);
+      return Just (maybe (empty) (append (x)) (m));
     }, Nothing, foldable);
   }
```

It's very similar to the change we made to remove the `isNothing` check from `init`.

File: index.js (hunk 19 above)
Comment (syves) [reply index.js]: I understand. There should be a conversation about why the change is or isn't an improvement.

Comment (syves) [reply index.js]: 🆗

File: index.js (hunk 20 above)
Comment (syves) [reply index.js]: This is very elegant and easier to read.

File: index.js (hunk 19 above)
Comment (davidchambers) [reply index.js]: Let's drop _of_ as well.

File: index.js (hunk 20 above)
Comment (syves) [reply index.js]: 🆗

**Comment by davidchambers** [comment]
I just added fast paths for arrays, so these functions will continue to operate efficiently on arrays. @syves, if you're happy with these changes please squash the commits in preparation for merging. :)

**Comment by davidchambers** [comment]
Thank you, @syves!
